File: creational-design-pattern/builder-pattern/builder.ts
/**
 * Builder Pattern
 * ================
 * Category: Creational Design Pattern
 *
 * Intent:
 *   Separate the construction of a complex object from its representation,
 *   allowing step-by-step construction with a fluent, chainable API.
 *
 * When to use:
 *   - When a constructor would need many parameters
 *   - When some parameters are optional and you want readable construction
 *   - When you want validation at build-time, not scattered across setters
 *
 * Key Participants:
 *   - Builder: Provides fluent methods for step-by-step construction
 *   - Product: The complex object being built (immutable once created)
 *   - Director (optional): Defines preset construction sequences
 */

// ---------------------------------------------------------------------------
// Product — immutable once built (all fields readonly)
// ---------------------------------------------------------------------------
interface HttpRequest {
  readonly method: string;
  readonly url: string;
  readonly headers: Record<string, string>;
  readonly queryParams: Record<string, string>;
  readonly body: string | null;
  readonly timeoutSeconds: number;
  readonly retries: number;
  readonly followRedirects: boolean;
}

function formatRequest(req: HttpRequest): string {
  let line = `${req.method} ${req.url}`;
  const qs = Object.entries(req.queryParams)
    .map(([k, v]) => `${k}=${v}`)
    .join("&");
  if (qs) line += `?${qs}`;

  const lines = [line];
  for (const [k, v] of Object.entries(req.headers)) {
    lines.push(`  ${k}: ${v}`);
  }
  if (req.body) {
    lines.push(`  Body: ${req.body.slice(0, 80)}...`);
  }
  lines.push(
    `  Timeout: ${req.timeoutSeconds}s | Retries: ${req.retries} | Follow redirects: ${req.followRedirects}`
  );
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// Builder — fluent API with method chaining
// ---------------------------------------------------------------------------
class HttpRequestBuilder {
  private _method = "GET";
  private _url = "";
  private _headers: Record<string, string> = {};
  private _queryParams: Record<string, string> = {};
  private _body: string | null = null;
  private _timeout = 30;
  private _retries = 0;
  private _followRedirects = true;

  // HTTP method shortcuts
  get(url: string): this {
    this._method = "GET";
    this._url = url;
    return this;
  }

  post(url: string): this {
    this._method = "POST";
    this._url = url;
    return this;
  }

  put(url: string): this {
    this._method = "PUT";
    this._url = url;
    return this;
  }

  delete(url: string): this {
    this._method = "DELETE";
    this._url = url;
    return this;
  }

  // Configuration methods (each returns `this` for chaining)
  header(key: string, value: string): this {
    this._headers[key] = value;
    return this;
  }

  query(key: string, value: string): this {
    this._queryParams[key] = value;
    return this;
  }

  body(content: string): this {
    this._body = content;
    return this;
  }

  timeout(seconds: number): this {
    this._timeout = seconds;
    return this;
  }

  retries(count: number): this {
    this._retries = count;
    return this;
  }

  noFollowRedirects(): this {
    this._followRedirects = false;
    return this;
  }

  // Build: validate and produce the immutable product
  build(): HttpRequest {
    if (!this._url) {
      throw new Error("URL is required");
    }
    if (["POST", "PUT"].includes(this._method) && this._body === null) {
      throw new Error(`${this._method} requests should have a body`);
    }

    return {
      method: this._method,
      url: this._url,
      headers: { ...this._headers },
      queryParams: { ...this._queryParams },
      body: this._body,
      timeoutSeconds: this._timeout,
      retries: this._retries,
      followRedirects: this._followRedirects,
    };
  }
}

// ---------------------------------------------------------------------------
// Director (optional) — preset construction sequences
// ---------------------------------------------------------------------------
class RequestDirector {
  static jsonApiGet(url: string): HttpRequestBuilder {
    return new HttpRequestBuilder()
      .get(url)
      .header("Accept", "application/json")
      .header("Content-Type", "application/json")
      .timeout(10)
      .retries(3);
  }

  static authenticatedPost(
    url: string,
    token: string,
    body: string
  ): HttpRequestBuilder {
    return new HttpRequestBuilder()
      .post(url)
      .header("Authorization", `Bearer ${token}`)
      .header("Content-Type", "application/json")
      .body(body)
      .timeout(30);
  }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
const request = new HttpRequestBuilder()
  .get("https://api.example.com/users")
  .header("Authorization", "Bearer abc123")
  .query("page", "1")
  .query("limit", "50")
  .timeout(15)
  .retries(2)
  .build();

console.log("Manual build:");
console.log(formatRequest(request));

console.log("\nDirector preset (JSON API GET):");
const apiRequest = RequestDirector.jsonApiGet(
  "https://api.example.com/products"
)
  .query("category", "electronics")
  .build();
console.log(formatRequest(apiRequest));

console.log("\nDirector preset (Authenticated POST):");
const postRequest = RequestDirector.authenticatedPost(
  "https://api.example.com/orders",
  "secret-token",
  '{"item": "laptop", "qty": 1}'
).build();
console.log(formatRequest(postRequest));

File: creational-design-pattern/builder-pattern/builder.py
"""
Builder Pattern
================
Category: Creational Design Pattern

Intent:
    Separate the construction of a complex object from its representation,
    allowing the same construction process to create different representations.
    Especially useful when an object has many optional parameters.

When to use:
    - When a constructor would need 5+ parameters (telescoping constructor problem)
    - When you want to create different representations of the same product
    - When construction involves multiple steps that should be chainable
    - When some parameters are optional and you want clear, readable construction

Key Participants:
    - Builder: Provides a fluent API with method chaining for step-by-step construction
    - Product: The complex object being built
    - Director (optional): Defines preset construction sequences

Why Builder over constructor with many params:
    - Named methods are self-documenting (`.with_timeout(30)` vs positional arg `30`)
    - Optional params don't require None/null placeholders
    - Validation can happen at build time, not scattered across setters
"""

from dataclasses import dataclass, field
from typing import Self


# ---------------------------------------------------------------------------
# Product — the complex object we're building
# ---------------------------------------------------------------------------
@dataclass(frozen=True)  # Immutable once built
class HttpRequest:
    """Represents a fully configured HTTP request."""
    method: str
    url: str
    headers: dict[str, str]
    query_params: dict[str, str]
    body: str | None
    timeout_seconds: int
    retries: int
    follow_redirects: bool

    def __str__(self) -> str:
        parts = [f"{self.method} {self.url}"]
        if self.query_params:
            qs = "&".join(f"{k}={v}" for k, v in self.query_params.items())
            parts[0] += f"?{qs}"
        for k, v in self.headers.items():
            parts.append(f"  {k}: {v}")
        if self.body:
            parts.append(f"  Body: {self.body[:80]}...")
        parts.append(f"  Timeout: {self.timeout_seconds}s | Retries: {self.retries} | Follow redirects: {self.follow_redirects}")
        return "\n".join(parts)


# ---------------------------------------------------------------------------
# Builder — fluent API with method chaining
# ---------------------------------------------------------------------------
class HttpRequestBuilder:
    """
    Builds an HttpRequest step by step. Each method returns `self`
    to enable chaining: HttpRequestBuilder().get(url).header(...).build()
    """

    def __init__(self) -> None:
        self._method: str = "GET"
        self._url: str = ""
        self._headers: dict[str, str] = {}
        self._query_params: dict[str, str] = {}
        self._body: str | None = None
        self._timeout: int = 30
        self._retries: int = 0
        self._follow_redirects: bool = True

    # --- HTTP method shortcuts ---
    def get(self, url: str) -> Self:
        self._method = "GET"
        self._url = url
        return self

    def post(self, url: str) -> Self:
        self._method = "POST"
        self._url = url
        return self

    def put(self, url: str) -> Self:
        self._method = "PUT"
        self._url = url
        return self

    def delete(self, url: str) -> Self:
        self._method = "DELETE"
        self._url = url
        return self

    # --- Configuration methods (each returns self for chaining) ---
    def header(self, key: str, value: str) -> Self:
        self._headers[key] = value
        return self

    def query(self, key: str, value: str) -> Self:
        self._query_params[key] = value
        return self

    def body(self, content: str) -> Self:
        self._body = content
        return self

    def timeout(self, seconds: int) -> Self:
        self._timeout = seconds
        return self

    def retries(self, count: int) -> Self:
        self._retries = count
        return self

    def no_follow_redirects(self) -> Self:
        self._follow_redirects = False
        return self

    # --- Build: validate and produce the immutable product ---
    def build(self) -> HttpRequest:
        if not self._url:
            raise ValueError("URL is required")
        if self._method in ("POST", "PUT") and self._body is None:
            raise ValueError(f"{self._method} requests should have a body")
        if self._timeout < 0:
            raise ValueError("Timeout must be non-negative")

        return HttpRequest(
            method=self._method,
            url=self._url,
            headers=dict(self._headers),
            query_params=dict(self._query_params),
            body=self._body,
            timeout_seconds=self._timeout,
            retries=self._retries,
            follow_redirects=self._follow_redirects,
        )


# ---------------------------------------------------------------------------
# Director (optional) — presets for common request configurations
# ---------------------------------------------------------------------------
class RequestDirector:
    """Provides pre-configured builder sequences for common use cases."""

    @staticmethod
    def json_api_get(url: str) -> HttpRequestBuilder:
        return (
            HttpRequestBuilder()
            .get(url)
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .timeout(10)
            .retries(3)
        )

    @staticmethod
    def authenticated_post(url: str, token: str, body: str) -> HttpRequestBuilder:
        return (
            HttpRequestBuilder()
            .post(url)
            .header("Authorization", f"Bearer {token}")
            .header("Content-Type", "application/json")
            .body(body)
            .timeout(30)
        )


# ---------------------------------------------------------------------------
# Usage Example
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # Manual step-by-step building with chaining
    request = (
        HttpRequestBuilder()
        .get("https://api.example.com/users")
        .header("Authorization", "Bearer abc123")
        .query("page", "1")
        .query("limit", "50")
        .timeout(15)
        .retries(2)
        .build()
    )
    print("Manual build:")
    print(request)

    # Using the Director for a preset configuration
    print("\nDirector preset (JSON API GET):")
    api_request = (
        RequestDirector.json_api_get("https://api.example.com/products")
        .query("category", "electronics")
        .build()
    )
    print(api_request)

    print("\nDirector preset (Authenticated POST):")
    post_request = (
        RequestDirector.authenticated_post(
            "https://api.example.com/orders",
            "secret-token",
            '{"item": "laptop", "qty": 1}',
        )
        .build()
    )
    print(post_request)
